File: sdks/typescript/src/transport.ts
import { StarfishError, type StarfishFrame, type DeliveryOptions } from "./types.js";

export interface RTCState {
  isPeerConnected(peerId: string): boolean;
  getConnectedPeerIds(): string[];
  getTopicPeers(topic: string): string[];
}

export type TransportDecision = { transport: "ws" } | { transport: "rtc"; peers: string[] };

export function selectTransport(
  frame: StarfishFrame,
  delivery: DeliveryOptions | undefined,
  rtcState: RTCState | null,
): TransportDecision {
  const prefer = delivery?.preferTransport ?? "auto";

  if (prefer === "ws") {
    return { transport: "ws" };
  }

  if (prefer === "rtc") {
    return selectRTC(frame, delivery, rtcState);
  }

  // auto
  return selectAuto(frame, delivery, rtcState);
}

function selectRTC(
  frame: StarfishFrame,
  delivery: DeliveryOptions | undefined,
  rtcState: RTCState | null,
): TransportDecision {
  const peers = resolveAvailablePeers(frame, rtcState);

  if (peers.length > 0) {
    return { transport: "rtc", peers };
  }

  if (delivery?.fallback !== false) {
    return { transport: "ws" };
  }

  throw new StarfishError(
    "TRANSPORT_UNAVAILABLE",
    "RTC transport is not available and fallback is disabled",
  );
}

function selectAuto(
  frame: StarfishFrame,
  delivery: DeliveryOptions | undefined,
  rtcState: RTCState | null,
): TransportDecision {
  const { resource, method } = frame.header;

  // data, session, presence → always WS
  if (resource === "data" || resource === "session" || resource === "presence") {
    return { transport: "ws" };
  }

  // topic publish
  if (resource === "topic" && method === "publish") {
    const reliability = delivery?.reliability ?? "reliable";
    if (reliability === "reliable") {
      return { transport: "ws" };
    }
    // unreliable/latest → RTC if peer path exists
    const topic = frame.header.topic;
    const peers = topic ? (rtcState?.getTopicPeers(topic) ?? []) : [];
    const connectedPeers = peers.filter((p) => rtcState?.isPeerConnected(p) ?? false);
    return connectedPeers.length > 0
      ? { transport: "rtc", peers: connectedPeers }
      : { transport: "ws" };
  }

  // message send
  if (resource === "message" && method === "send") {
    const reliability = delivery?.reliability ?? "reliable";
    const peers = resolveAvailablePeers(frame, rtcState);

    if (reliability === "reliable") {
      // RTC if connected, else WS
      return peers.length > 0 ? { transport: "rtc", peers } : { transport: "ws" };
    }

    // unreliable/latest → RTC preferred
    return peers.length > 0 ? { transport: "rtc", peers } : { transport: "ws" };
  }

  // Everything else → WS
  return { transport: "ws" };
}

/** Resolve peers from frame.header.to (direct messaging) or topic peers (topic publishing) */
function resolveAvailablePeers(frame: StarfishFrame, rtcState: RTCState | null): string[] {
  if (!rtcState) return [];

  // For topic publish, resolve peers from the topic subscription map
  if (
    frame.header.resource === "topic" &&
    frame.header.method === "publish" &&
    frame.header.topic
  ) {
    const topicPeers = rtcState.getTopicPeers(frame.header.topic);
    return topicPeers.filter((peerId) => rtcState.isPeerConnected(peerId));
  }

  // For direct messages, resolve from frame.header.to
  const to = frame.header.to;
  const targets = Array.isArray(to) ? to : to ? [to] : [];

  return targets.filter((peerId) => rtcState.isPeerConnected(peerId));
}

File: sdks/swift/Sources/StarfishClient/Transport.swift
import Foundation

/// Result of transport selection.
enum TransportDecision: Sendable, Equatable {
    case ws
    case rtc(peers: [String])
}

/// Protocol for querying RTC peer state (used by transport selection).
protocol RTCState {
    func isPeerConnected(_ peerId: String) -> Bool
    func getConnectedPeerIds() -> [String]
    func getTopicPeers(_ topic: String) -> [String]
}

/// Selects the appropriate transport (WS or RTC) for a given frame.
func selectTransport(
    frame: StarfishFrame,
    delivery: DeliveryOptions?,
    rtcState: RTCState?
) throws -> TransportDecision {
    let prefer = delivery?.preferTransport ?? .auto

    switch prefer {
    case .ws:
        return .ws
    case .rtc:
        return try selectRTC(frame: frame, delivery: delivery, rtcState: rtcState)
    case .auto:
        return selectAuto(frame: frame, delivery: delivery, rtcState: rtcState)
    }
}

private func selectRTC(
    frame: StarfishFrame,
    delivery: DeliveryOptions?,
    rtcState: RTCState?
) throws -> TransportDecision {
    let peers = resolveAvailablePeers(frame: frame, rtcState: rtcState)

    if !peers.isEmpty {
        return .rtc(peers: peers)
    }

    if delivery?.fallback != false {
        return .ws
    }

    throw StarfishError(
        code: .transportUnavailable,
        message: "RTC transport is not available and fallback is disabled"
    )
}

private func selectAuto(
    frame: StarfishFrame,
    delivery: DeliveryOptions?,
    rtcState: RTCState?
) -> TransportDecision {
    let resource = frame.header.resource
    let method = frame.header.method

    // data.*, session.*, presence.* -> always WS
    if resource == "data" || resource == "session" || resource == "presence" {
        return .ws
    }

    // topic.publish
    if resource == "topic" && method == "publish" {
        let reliability = delivery?.reliability ?? .reliable
        if reliability == .reliable {
            return .ws
        }
        // unreliable/latest -> RTC if peer path exists
        guard let topic = frame.header.topic, let rtcState = rtcState else { return .ws }
        let peers = rtcState.getTopicPeers(topic)
        let connectedPeers = peers.filter { rtcState.isPeerConnected($0) }
        return connectedPeers.isEmpty ? .ws : .rtc(peers: connectedPeers)
    }

    // message.send
    if resource == "message" && method == "send" {
        let peers = resolveAvailablePeers(frame: frame, rtcState: rtcState)
        return peers.isEmpty ? .ws : .rtc(peers: peers)
    }

    // Everything else -> WS
    return .ws
}

private func resolveAvailablePeers(frame: StarfishFrame, rtcState: RTCState?) -> [String] {
    guard let rtcState = rtcState else { return [] }

    // For topic.publish, resolve from topic subscription map
    if frame.header.resource == "topic" && frame.header.method == "publish", let topic = frame.header.topic {
        return rtcState.getTopicPeers(topic).filter { rtcState.isPeerConnected($0) }
    }

    // For direct messages, resolve from frame.to
    let targets: [String]
    switch frame.header.to {
    case .single(let id):
        targets = [id]
    case .multiple(let ids):
        targets = ids
    case nil:
        targets = []
    }

    return targets.filter { rtcState.isPeerConnected($0) }
}
